src/parser.rs

```diff
@@ -407,6 +411,243 @@ pub fn parse_gcov(gcov_path: &Path) -> Result<Vec<(String,CovResult)>, ParserErr
+                            hash_map::Entry::Vacant(v) => {
+                                v.insert(CovResult {
+                                    functions,
+                                    lines: BTreeMap::new(),
+                                    branches: BTreeMap::new(),
+                                });
+                            },
+                        };
+                    },
+                    "sourcefile" => {
+                        // Find source file name
+                        let sourcefile = get_xml_attribute(&attributes, "name").unwrap(); // Person.java
+                        let class = sourcefile.trim_right_matches(".java"); // Person
+
+                        // Process all <line /> and <counter /> for this class
+                        let (lines, branches) = parse_java_sourcefile(parser).unwrap();
+
+                        // Save the result
+                        match results_hash.entry(class.to_string()) {
+                            hash_map::Entry::Occupied(obj) => {
+                                let obj = obj.into_mut();
+                                obj.lines = lines;
+                                obj.branches = branches;
+                            },
+                            hash_map::Entry::Vacant(v) => {
+                                v.insert(CovResult {
+                                    functions: HashMap::new(),
+                                    lines,
+                                    branches,
+                                });
+                            },
+                        };
+                    },
+                    &_ => {},
+                }
+            },
+            Ok(XmlEvent::EndElement{name}) => {
+                match name.local_name.as_str() {
+                    "package" => break,
+                    &_ => {}
+                }
+            },
+            Err(e) => {
+                return Err(ParserError::Parse(e.to_string()))
+            },
+            _ => {}
+        }
+    }
+
+    Ok(
+        results_hash.into_iter().filter(
+            |(_, result)| result.lines.len() > 0 || result.branches.len() > 0
+        ).map(
+            |(class, value)| (format!("{}/{}.java", package, class).trim_left_matches('/').to_string(), value)
+        ).collect()
+    )
+}
+
+pub fn parse_jacoco_xml<T: Read>(xml_reader: BufReader<T>) -> Result<Vec<(String,CovResult)>, ParserError> {
+
+    let mut parser = EventReader::new(xml_reader);
+    let mut results = Vec::new();
+
+    loop {
+        match parser.next() {
+            Ok(XmlEvent::StartDocument { .. }) => {},
+            Ok(XmlEvent::EndDocument) => break,
+            Ok(XmlEvent::StartElement { name, attributes, .. }) => {
+                match name.local_name.as_str() {
+                    "package" => {
+                        // Find package name
+                        let package = get_xml_attribute(&attributes, "name").unwrap();
```

Comment:
Let's use `except` here, so we can print a human-readable message (or we could return a ParserError, but since this is really unexpected I'd just use `except`).

Reply:
I could make `get_xml_attribute` to return `Result<>` and use `?` everywhere to propagate all the errors.

More or less, remove all `unwrap()` calls?

Reply:
Yes, that would work!

Reply:
So, use `?` when it's an error that could happen, and `expect` or `unwrap` when it something that is almost impossible.